Fix ScanForWalletTransactions to return an enum indicating scan result: success / failure / user_abort

Empact opened the pull request:
Return the failed block as an out arg.

Fixes #11450.

/cc #12275

Empact committed:
Make CWallet::ScanForWalletTransactions args and return value const

Empact commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3767,16 +3767,13 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
+    } else if (pwallet->IsAbortingRescan()) {
+        throw JSONRPCError(RPC_MISC_ERROR, "Rescan aborted.");
+    } else {
+        throw JSONRPCError(RPC_MISC_ERROR, strprintf("Rescan failed. Potentially corrupted data files. Failed at height %i.", stopBlock->nHeight));
```
Note I added the failure height here. Happy to remove, but I assume it could be helpful.

Empact commented on src/wallet/wallet.cpp:
```diff
@@ -4167,11 +4167,11 @@ CWallet* CWallet::CreateWalletFromFile(const std::string& name, const fs::path&
             WalletRescanReserver reserver(walletInstance);
-            if (!reserver.reserve()) {
+            const CBlockIndex* stop_block = nullptr;
+            if (!reserver.reserve() || !walletInstance->ScanForWalletTransactions(pindexRescan, nullptr, stop_block, reserver, true)) {
```
Note this did not previously fail when ScanForWalletTransactions failed, so this removes a silent failure.

promag replied:
I think this is ok. This just adds another case where `walletInstance` is leaked on error.

Empact replied:
Should we be switching `walletInstance` to RAII then?

Empact commented on src/wallet/wallet.cpp:
```diff
@@ -1791,7 +1791,7 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, CBlock
         ShowProgress(_("Rescanning..."), 100); // hide progress dialog in GUI
     }
-    return ret;
+    return (!fAbortRescan && !failed_block);
```
Alternatively, this could be `(!pindex && !failed_block)`

jonasschnelli commented on src/wallet/wallet.cpp:
```diff
@@ -1708,9 +1709,10 @@ int64_t CWallet::RescanFromTime(int64_t startTime, const WalletRescanReserver& r
+ * Returns true if scan was successful. If a complete rescan was not
+ * possible (due to pruning or corruption), returns false and sets failed_block
+ * to the most recent block that could not be scanned. If the rescan was aborted
+ * before it could complete, returns false and does not set failed_block.
```
Not sure if in this PR, but would it make sense to just return an `enum` to better distinct between `fail`, `ok` and `user-aborted`?

promag replied:
Could be done here IMO, not a big change and would be a lot better IMO.

Empact replied:
Updated to return an enum.

promag commented on src/wallet/wallet.cpp:
```diff
@@ -1719,7 +1721,7 @@ int64_t CWallet::RescanFromTime(int64_t startTime, const WalletRescanReserver& r
  * transactions for.
  */
-CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, CBlockIndex* pindexStop, const WalletRescanReserver &reserver, bool fUpdate)
+bool CWallet::ScanForWalletTransactions(const CBlockIndex* const pindexStart, const CBlockIndex* const pindexStop, const CBlockIndex*& failed_block, const WalletRescanReserver& reserver, bool fUpdate)
```
*Output* arguments could be last?

Empact replied:
Made it second-to-last due to default value for update arg.

promag commented on src/qt/test/wallettests.cpp:
```diff
@@ -156,7 +156,8 @@ void TestGUI()
         reserver.reserve();
-        wallet.ScanForWalletTransactions(chainActive.Genesis(), nullptr, reserver, true);
+        const CBlockIndex* stop_block = nullptr;
+        wallet.ScanForWalletTransactions(chainActive.Genesis(), nullptr, stop_block, reserver, true);
```
Could assert return value and `stop_block` value?

promag reviewed (commented):
Concept ACK.

Empact committed:
Return a status enum from ScanForWalletTransactions

Return the failed block as an out var.

This clarifies the outcome as the prior return value could
be null due to user abort or failure.

Empact commented on src/wallet/wallet.cpp:
```diff
@@ -1763,14 +1764,14 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, CBlock
                     // Abort scan if current block is no longer active, to prevent
                     // marking transactions as coming from the wrong block.
-                    ret = pindex;
+                    failed_block = pindex;
```
Not sure about that - my inclination is to maintain stricter checking at the risk of spurious failures over less strict checking at the risk of silent failure. How about opening a separate PR for that so it can be considered separately?

Empact commented on src/wallet/wallet.cpp:
```diff
@@ -1695,8 +1695,9 @@ int64_t CWallet::RescanFromTime(int64_t startTime, const WalletRescanReserver& r
-        const CBlockIndex* const failedBlock = ScanForWalletTransactions(startBlock, nullptr, reserver, update);
-        if (failedBlock) {
+        const CBlockIndex* failedBlock = nullptr;
+        // TODO: this should take into account failure by ScanResult::USER_ABORTED
```
I'll address this todo in a follow-up.

Empact replied:
Follow-up is here, I'll wait to open until this is accepted:
https://github.com/Empact/bitcoin/tree/rescan-from-time

Empact replied:
#14942

jonasschnelli commented on src/wallet/wallet.h:
```diff
@@ -914,7 +914,13 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
-    CBlockIndex* ScanForWalletTransactions(CBlockIndex* pindexStart, CBlockIndex* pindexStop, const WalletRescanReserver& reserver, bool fUpdate = false);
+
+    enum class ScanResult : int {
+        SUCCESS = 0,
```
nit: not sure if SUCCESS should use a non C-ish, non null positive value (1 probably)

MarcoFalke replied:
note that there is no need for an `enum class` to derive from int or even specify values for the members.

Empact commented:
Dropped explicit values and type from `ScanResult` enum.

Empact commented:
Rebased and updated to accommodate #12507.

Empact commented:
Rebased and moved the shutdown / abort handling inline.

sabran125 reviewed (commented):
indicating the requested changes

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3912,16 +3912,19 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
+    case CWallet::ScanResult::USER_ABORT:
+        throw JSONRPCError(RPC_MISC_ERROR, "Rescan aborted.");
+    default:
+        throw JSONRPCError(RPC_MISC_ERROR, strprintf("Unexpected rescan result: %i.", static_cast<int>(result)));
```
Should remove the default case, so instead of a runtime-error the compiler can warn about missing cases.
```suggestion
        // no default case, so the compiler can warn about missing cases
```

maflcko reviewed (commented):
Concept ACK. Needs rebase

Empact commented:
Rebased. @MarcoFalke could you remove "Up for grabs"?

MarcoFalke commented on src/qt/test/wallettests.cpp:
```diff
@@ -144,7 +144,12 @@ void TestGUI()
         reserver.reserve();
-        wallet->ScanForWalletTransactions(chainActive.Genesis(), nullptr, reserver, true);
+        const CBlockIndex* const null_block = nullptr;
+        const CBlockIndex* stop_block = nullptr;
```
nit: No need to initialize a return value. In fact this might look like you are passing in something.

Also could inline nullptr for null_block below for brevity.

Empact replied:
Removed the initialization of the out var, though this required setting the out var to null in the method for consistent behavior under test. Seems like a good change to me as this seems more predictable.

Can't inline the `null_block` vars as that results in template ambiguity around `<ostream, nullptr_t>`

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3297,16 +3297,18 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
-            throw JSONRPCError(RPC_MISC_ERROR, "Rescan aborted.");
-        }
-        // if we got a nullptr returned, ScanForWalletTransactions did rescan up to the requested stopindex
+    const CBlockIndex* stopBlock = nullptr;
```
nit: No need to initialize a return value. In fact this might look like you are passing in something.

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3297,16 +3297,18 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
+        pwallet->ScanForWalletTransactions(pindexStart, pindexStop, reserver, stopBlock, true);
+    switch (result) {
+    case CWallet::ScanResult::SUCCESS:
         stopBlock = pindexStop ? pindexStop : pChainTip;
```
Why are you overriding the return value?

Empact replied:
This has been the behavior since the rpc's introduction in c77170fbdb

MarcoFalke replied:
I can pick this up and address my feedback middle of next week. (Let me know if you are currently busy or want to take a stab at this.)

MarcoFalke commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -50,7 +50,9 @@ BOOST_FIXTURE_TEST_CASE(rescan, TestChain100Setup)
-        BOOST_CHECK_EQUAL(nullBlock, wallet.ScanForWalletTransactions(oldTip, nullptr, reserver));
+        const CBlockIndex* stop_block = nullptr;
+        BOOST_CHECK_EQUAL(wallet.ScanForWalletTransactions(oldTip, nullptr, reserver, stop_block), CWallet::ScanResult::SUCCESS);
+        BOOST_CHECK_EQUAL(nullBlock, stop_block);
```
pIndexStop was nullptr, so this should be tip instead?

MeshCollider replied:
@MarcoFalke `stop_block` is only set if the scan failed, it isn't set to tip if the scan succeeds

MarcoFalke commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -279,7 +283,10 @@ class ListCoinsTestingSetup : public TestChain100Setup
+        const CBlockIndex* const null_block = nullptr;
+        const CBlockIndex* stop_block = nullptr;
+        BOOST_CHECK_EQUAL(wallet->ScanForWalletTransactions(chainActive.Genesis(), nullptr, reserver, stop_block), CWallet::ScanResult::SUCCESS);
+        BOOST_CHECK_EQUAL(stop_block, null_block);
```
Should be tip instead of null_block?

MeshCollider replied:
Same here as the other case :)

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -1619,18 +1620,20 @@ int64_t CWallet::RescanFromTime(int64_t startTime, const WalletRescanReserver& r
- * block that could not be scanned.
+ * @param[in] pindexStop if not a nullptr, the scan will stop at this block-index
+ * @param[out] failed_block if FAILURE is returned, will be set to the most
+ *    recent block that could not be scanned.
```
nit: could return `last_block`, which will either be "the most recent block that could not be scanned" (failure/abort) or "pindexStop/::chainActive.Tip()" (success)

maflcko reviewed (approved):
Concept ACK, but I'd prefer it the block was returned in the case of success as well. (Otherwise it is up to the caller to *guess* which was the last successfully scanned block, since the tip can change without holding cs_main at the call site)

MeshCollider commented:
utACK https://github.com/bitcoin/bitcoin/pull/13076/commits/b52abc7164ba6403124e409af2e3ee2e2dad7f24

I agree with Marco that returning the last successful block is probably a good idea, but with that and the other comments, at least this doesn't appear to make things worse, so it could be addressed in follow-ups if not here :)

ryanofsky commented:
> I agree with Marco that returning the last successful block is probably a good idea

Returning the last failed block instead of the last successful block is needed for importmulti. Importmulti uses the last failed block to return success for keys with birthdays after that block, and failures for keys with older birthdays.

I tried to clarify this in #12275 by returning both values (`last_scanned` and `last_failed`):

```c++
    struct ScanResult
    {
        const CBlockIndex* first_scanned = nullptr;
        const CBlockIndex* last_scanned = nullptr;
        const CBlockIndex* first_failed = nullptr;
        const CBlockIndex* last_failed = nullptr;
    };
```

Which I still think would be an improvement over the current code.

MeshCollider commented:
@ryanofsky I think I mis-explained, I mean in the case of success - the failure case behaviour should remain unchanged

Empact committed:
Add stop_block out arg to ScanForWalletTransactions

Accurately reports the last block successfully scanned, replacing a return of
the chain tip, which represented possibly inaccurated data in a race condition.

MarcoFalke commented on src/qt/test/wallettests.cpp:
```diff
@@ -146,7 +146,13 @@ void TestGUI()
+        QCOMPARE(
+            wallet->ScanForWalletTransactions(chainActive.Genesis(), nullptr, reserver, failed_block, stop_block, true /* fUpdate */),
+            CWallet::ScanResult::SUCCESS);
+        QCOMPARE(stop_block, null_block);
```
Travis failure:

```
********* Start testing of WalletTests *********
Config: Using QtTest library 5.9.6, Qt 5.9.6 (i386-little_endian-ilp32 static release build; by GCC 7.3.0)
PASS   : WalletTests::initTestCase()
FAIL!  : WalletTests::walletTests() Compared pointers are not the same
   Loc: [qt/test/wallettests.cpp(154)]
PASS   : WalletTests::cleanupTestCase()
Totals: 2 passed, 1 failed, 0 skipped, 0 blacklisted, 245ms
********* Finished testing of WalletTests *********

Empact replied:
Thanks! Solved.

ryanofsky reviewed (approved):
utACK bd3b0361d840bff95988a048abf70ade94d80524

jonasschnelli commented:
utACK bd3b0361d840bff95988a048abf70ade94d80524

MarcoFalke commented:
utACK bd3b0361d8

MeshCollider commented:
utACK https://github.com/bitcoin/bitcoin/pull/13076/commits/bd3b0361d840bff95988a048abf70ade94d80524

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1647,8 +1652,8 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, CBlock
-    CBlockIndex* ret = nullptr;
+    const CBlockIndex* pindex = pindexStart;
+    failed_block = nullptr;
 
```
Code is not setting `stop_block` null here, which seems unintended and could lead to uninitialized variables. Would be good to fix this or add documentation if it was done intentionally.

Empact replied:
Ah, right, the code does contradict the comments on line 1634. How about I fix this in #14942?
